src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -185,121 +166,189 @@ def __enter__(self):
         self._processes_manager.start_processes()
         self._processes_manager.ensure_healthy()
 
-        monitor_pool = ThreadPool(self._n_process, initializer=set_context, initargs=(contextvars.copy_context(),))
-        self._monitor_pool = monitor_pool
-
-        return self
-
-    def __exit__(self, exc_type, exc_val, exc_tb):
+        # Start a thread pool to monitor the processes.
+        self._monitor_pool = ThreadPool(
+            self._n_process, initializer=set_context, initargs=(contextvars.copy_context(),)
+        )
+        # The variable '_async_tasks' here is a list of AsyncResult object
+        # that can be used to check if the execution are finished.
+        # The actual line results of the batch run are stored in 'result_dict'.
+
+        # Create _n_process monitoring threads, mainly used to assign tasks and receive line result.
+        # When receiving terminate signal, end the process.
+        # When line execution timeout or process crash, restart the process.
+        self._async_tasks = [
+            self._monitor_pool.apply_async(
+                func=self._monitor_workers_and_process_tasks_in_thread,
+                args=(
+                    self._task_queue,  # Shared task queue for all sub processes to read the input data.
+                    self._result_dict,  # Line result dict of the batch run.
+                    i,  # Index of the sub process.
+                    self._input_queues[i],  # Specific input queue for sub process, used to send input data to it.
+                    self._output_queues[i],  # Specific output queue for sub process, used to receive results from it.
+                ),
+            )
+            for i in range(self._n_process)
+        ]
+
+    def close(self):
```

Comment:
Iterate self._process info. If not empty, force kill the process.

Reply:
Added function to process manager.